pootle/apps/accounts/utils.py

```diff
@@ -30,3 +33,215 @@ def get_user_by_email(email):
+    - units: submitted_by, commented_by, reviewed_by
+    - submissions: submitter
+    - suggestions: user, reviewer
+
+    :param src_user: `User` instance to merge from.
+    :param target_user: `User` instance to merge to.
+    """
+
+    # Update submitted_by, commented_by and reviewed_by on units
+    for unit in src_user.submitted.iterator():
+        unit.submitted_by = target_user
+        unit.save()
+
+    for unit in src_user.commented.iterator():
+        unit.commented_by = target_user
+        unit.save()
+
+    for unit in src_user.reviewed.iterator():
+        unit.reviewed_by = target_user
+        unit.save()
+
+    # Update submitter on submissions
+    for submission in src_user.submission_set.iterator():
+        submission.submitter = target_user
+
+        # Before we can save we first have to remove existing score_logs for
+        # this submission - they will be recreated on save with correct user.
+        for score_log in submission.scorelog_set.iterator():
+            score_log.delete()
+        submission.save()
+
+    # Update user and reviewer on suggestions
+    for suggestion in src_user.suggestions.iterator():
+        suggestion.user = target_user
+        suggestion.save()
+
+    for suggestion in src_user.reviews.iterator():
+        suggestion.reviewer = target_user
+        suggestion.save()
+
+
+def remove_units_created_by(user):
+    """Remove units created by user that have not had further
+    activity.
+
+    :param user: `User` to remove units for.
+    """
+
+    # Delete units created by user without submissions by others.
+    for unit in user.units_created.iterator():
+
+        # Find submissions by other users on this unit.
+        other_subs = unit.submission_set.exclude(submitter=user)
+
+        if not other_subs.exists():
+            unit.delete()
+
+
+def revert_units_state_changed_by(user):
+    """Revert unit edits made by a user to previous edit.
+
+    :param user: `User` to revert submissions for.
+    """
+    for submission in user.unit_states_changed.iterator():
+        unit = submission.unit
+
+        # We have to get latest by pk as on mysql precision is not to
+        # microseconds - so creation_time can be ambiguous
+        if submission != unit.state_changes.latest('pk'):
+            # If the unit has been changed more recently we don't need to
+            #  revert the unit state.
+            submission.delete()
+            return
+        submission.delete()
+        other_submissions = unit.state_changes.exclude(submitter=user)
+        if other_submissions.exists():
+            new_state = other_submissions.latest().new_value
+        else:
+            new_state = UNTRANSLATED
+        if new_state != unit.state:
+            if unit.state == FUZZY:
+                unit.markfuzzy(False)
+            elif new_state == FUZZY:
+                unit.markfuzzy(True)
+            unit.state = new_state
+            unit._state_updated = True
```

Comment:
We probably want to add a note on why this is necessary. Ditto on the other occurrences of `unit._*_updated`